File: src/main/java/me/wawwior/utils/hashing/hashes/DefaultHashProvider.java
package me.wawwior.utils.hashing.hashes;

import me.wawwior.utils.hashing.Hash;
import me.wawwior.utils.hashing.HashProvider;
import me.wawwior.utils.hashing.Salt;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

public class DefaultHashProvider implements HashProvider {

    private final MessageDigest digest;

    private final Hash instance;

    public DefaultHashProvider(String algorithm) {
        try {
            digest = MessageDigest.getInstance(algorithm);
            instance = new DefaultHash(digest);
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public Hash hash() {
        return instance;
    }

    @Override
    public Hash salted(Salt salt) {
        return new DefaultSaltedHash(digest, salt);
    }
}

File: src/main/kotlin/me/wawwior/utils/hashing/hashes/DefaultHashProvider.kt
package me.wawwior.utils.hashing.hashes

import me.wawwior.utils.hashing.HashProvider
import me.wawwior.utils.hashing.Salt
import java.security.MessageDigest

class DefaultHashProvider(algorithm: String) : HashProvider {

    private val messageDigest = MessageDigest.getInstance(algorithm)

    private val instance = DefaultHash(messageDigest)

    override fun hash() = instance

    override fun salted(salt: Salt) = DefaultSaltedHash(messageDigest, salt)
}